src/transfer.cpp

```diff
@@ -1199,4 +1252,431 @@ DirectReadSlot::~DirectReadSlot()
+        transfers[transfer->type].push_back(transfer);
+        client->transfercacheadd(transfer);
+    }
+    else
+    {
+        transfer_list::iterator it = std::lower_bound(transfers[transfer->type].begin(), transfers[transfer->type].end(), transfer, priority_comparator);
+        assert(it == transfers[transfer->type].end() || (*it)->priority != transfer->priority);
+        transfers[transfer->type].insert(it, transfer);
+    }
+}
+
+void TransferList::removetransfer(Transfer *transfer)
+{
+    transfer_list::iterator it = iterator(transfer);
+    if (it != transfers[transfer->type].end())
+    {
+        transfers[transfer->type].erase(it);
+    }
+}
+
+void TransferList::movetransfer(Transfer *transfer, Transfer *prevTransfer)
+{
+    transfer_list::iterator dstit = iterator(prevTransfer);
+    if (dstit == transfers[prevTransfer->type].end())
+    {
+        return;
+    }
+    movetransfer(transfer, dstit);
+}
+
+void TransferList::movetransfer(Transfer *transfer, unsigned int position)
+{
+    transfer_list::iterator it = iterator(transfer);
+    if (it == transfers[transfer->type].end())
+    {
+        return;
+    }
+
+    if (position >= transfers[transfer->type].size())
+    {
+        transfers[transfer->type].erase(it);
+        currentpriority += PRIORITY_STEP;
+        transfer->priority = currentpriority;
+        assert(!transfers[transfer->type].size() || transfers[transfer->type][transfers[transfer->type].size() - 1]->priority < transfer->priority);
+        transfers[transfer->type].push_back(transfer);
+        client->transfercacheadd(transfer);
+        client->app->transfer_update(transfer);
+        return;
+    }
+
+    transfer_list::iterator dstit = transfers[transfer->type].begin() + position;
+    movetransfer(it, dstit);
+}
+
+void TransferList::movetransfer(Transfer *transfer, transfer_list::iterator dstit)
+{
+    transfer_list::iterator it = iterator(transfer);
+    if (it == transfers[transfer->type].end())
+    {
+        return;
+    }
+    movetransfer(it, dstit);
+}
+
+void TransferList::movetransfer(transfer_list::iterator it, transfer_list::iterator dstit)
+{
+    if (it == dstit)
+    {
+        LOG_warn << "Trying to move to the same position";
+        return;
+    }
+
+    Transfer *transfer = (*it);
+    if (it == (transfers[transfer->type].end() - 1)
+            && dstit == transfers[transfer->type].end())
+    {
+        LOG_warn << "Trying to move the last transfer down";
+        return;
+    }
+
+    if (dstit == transfers[transfer->type].end()
+            || (dstit + 1) == transfers[transfer->type].end())
+    {
+        LOG_debug << "Moving transfer to the last position";
+        if (transfer->slot)
+        {
+            transfer_list::iterator cit = it + 1;
+            while (cit != transfers[transfer->type].end())
+            {
+                if (!(*cit)->slot)
```

Comment:
Check not only that the transfer doesn't have one slot, but also that is electable for dispatching.